### PR TITLE
Added web access request

### PR DESCRIPTION
**Pull request by leonardlan**
Added web access request feature:

Trainee view:
create, update, delete, and view forms
use time on request

TA view
TA approve, deny, mark for fellowship, comment

### CODE REVIEW

File: ap/web_access/forms.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+class WebAccessRequestCreateForm(forms.ModelForm):
+
+    date_expire = forms.DateField(widget=DateInput())
+
+    def clean_date_expire(self):
+        """ Invalid form if date expire is earlier than today """
+        data = self.cleaned_data['date_expire']
+        if datetime.now().date() > data:
+            raise forms.ValidationError("Date expire has already passed.")
```
Comment (radicalray) [inline ap/web_access/forms.py]: Is this catched in the UI?

File: ap/web_access/utils.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import re
+import socket
+import subprocess
+
+"""
+This module contains web-access request specific Python functions for
+communicating with the firewall.
+"""
+
+# TODO: Move these either to the database or wherever app specific settings should be
```
Comment (leonardlan) [inline ap/web_access/utils.py]: lol totally forgot about this.

File: ap/web_access/forms.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from django import forms
+
+from .models import WebRequest
+
+from functools import partial
+from datetime import datetime
+
+DateInput = partial(forms.DateInput, {'class': 'datepicker'})
```
Comment (leonardlan) [inline ap/web_access/forms.py]: Put this comment: "Needed for JQuery datepicker UI to work."

File: ap/web_access/forms.py (hunk 1 above)
Comment (leonardlan) [inline ap/web_access/forms.py]: When you click "Submit", an error will pop with the error and not allow request to be submitted. :dancers:

File: ap/web_access/models.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+from django.db import models
+from django.core.urlresolvers import reverse
+# from macaddress.fields import MACAddressField
```
Comment (leonardlan) [inline ap/web_access/models.py]: Remove

File: ap/web_access/views.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+class TAWebAccessUpdate(generic.UpdateView):
+
+    model = WebRequest
+    template_name = 'web_access/ta_web_access_update.html'
+    form_class = WebAccessRequestTACommentForm
+    context_object_name = 'web_access'
+
+
+def modify_status(request, status, id):
```
Comment (leonardlan) [inline ap/web_access/views.py]: Add this comment: """ Changes status of web access request """

File: ap/web_access/models.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    def get_update_url(self):
+        return reverse('web_access:web_access-update', kwargs={'pk': self.id})
+
+    def get_absolute_url(self):
+        return reverse('web_access:web_access-detail', kwargs={'pk': self.id})
+
+    # Sort by trainee name
+    class Meta:
+        ordering = ['trainee__account__firstname']
+
```
Comment (burricks) [inline ap/web_access/models.py]: Would it make more sense to sort by expiration date?

File: ap/web_access/templates/web_access/webrequest_list.html [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        </div>
+    {% empty %}
+        <h4><i>You have no web access requests.</i></h4>
+    {% endfor %}
+    </div>
+
+    <script type="text/javascript">
+        $(document).on('click', ".delete_request_button", function() {
+            var accessID = +$(this).val();
+            console.log(accessID);
```
Comment (burricks) [inline ap/web_access/templates/web_access/webrequest_list.html]: What is the motivation for logging the accessID on the console?

File: ap/web_access/models.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    # The date the request was started.
+    time_started = models.DateTimeField(auto_now_add=False, blank=True, null=True)
+
+    # The date the request expires
+    date_expire = models.DateField()
+
+    # For a guest web access request this is used to identify the request.
+    # for non guests this field is set when the web access request is started.
+    # hw_address = MACAddressField(blank=True)
```
Comment (radicalray) [inline ap/web_access/models.py]: Why was this taken out?

File: ap/web_access/models.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    date_assigned = models.DateTimeField(auto_now_add=True)
+
+    # The date the request was started.
+    time_started = models.DateTimeField(auto_now_add=False, blank=True, null=True)
+
+    # The date the request expires
+    date_expire = models.DateField()
+
+    # For a guest web access request this is used to identify the request.
+    # for non guests this field is set when the web access request is started.
```
Comment (radicalray) [inline ap/web_access/models.py]: Do we handle a "Guest Web Access Request"? Before they sign in, saints can request for internet from login screen?

File: ap/web_access/templates/web_access/web_access_create.html [hunk 10]
```diff
@@ -0,0 +1,31 @@
+{% extends "base.html" %}
+{% load staticfiles %}
+{% load bootstrap3 %}
+{% bootstrap_css %}
+{% bootstrap_javascript %}
+{% bootstrap_messages %}
+
+{% block content %}
```
Comment (radicalray) [inline ap/web_access/templates/web_access/web_access_create.html]: Use {% block scripts %} to put this in head. Don't put javascript in the body.

File: ap/web_access/templates/web_access/webrequest_list.html [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                </div>
+            </div>
+        {% endfor %}
+        </div>
+    {% empty %}
+        <h4><i>You have no web access requests.</i></h4>
+    {% endfor %}
+    </div>
+
+    <script type="text/javascript">
```
Comment (radicalray) [inline ap/web_access/templates/web_access/webrequest_list.html]: Put this in the $(document).ready and in <head> not in body.

File: ap/web_access/utils.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import socket
+import subprocess
+
+"""
+This module contains web-access request specific Python functions for
+communicating with the firewall.
+"""
+
+# TODO: Move these either to the database or wherever app specific settings should be 
+HOST = "10.0.8.20" # hostname or ip address of the firewall (add to /etc/hosts)
```
Comment (radicalray) [inline ap/web_access/utils.py]: Move HOST + PORT to prod.py and to dev.py and import the parameters into this file. This way, when we move servers, easy to find the settings to change.

File: requirements/base.txt [hunk 13]
```diff
@@ -31,4 +31,7 @@ wsgiref==0.1.2
 librabbitmq==1.6.1
 celery==3.1.19
 django-celery==3.1.17
-flower>=0.8.3
\ No newline at end of file
+flower>=0.8.3
```
Comment (radicalray) [inline requirements/base.txt]: Comment this to say it's for websockets for web-request.

File: ap/web_access/models.py (hunk 6 above)
Comment (leonardlan) [inline ap/web_access/models.py]: It's listed by expiration date on trainee view and TA view already

File: ap/web_access/templates/web_access/webrequest_list.html (hunk 7 above)
Comment (leonardlan) [inline ap/web_access/templates/web_access/webrequest_list.html]: lol. I think I copied code from leaveslips. I'll take it out.

File: ap/web_access/models.py (hunk 8 above)
Comment (leonardlan) [inline ap/web_access/models.py]: Alex Webb had it in his code. I have no idea what it does and am too lazy to figure it out. It is not used anywhere else so I'm going to remove it.

File: ap/web_access/models.py (hunk 9 above)
Comment (leonardlan) [inline ap/web_access/models.py]: future request.

File: ap/web_access/utils.py (hunk 12 above)
Comment (leonardlan) [inline ap/web_access/utils.py]: How do I import the parameters into this file? I don't see an example anywhere.

File: ap/web_access/models.py (hunk 6 above)
Comment (radicalray) [inline ap/web_access/models.py]: This definition is for admin view. Make it order by start date, expiration date and then by firstname categorical sort.

File: ap/web_access/models.py (hunk 8 above)
Comment (radicalray) [inline ap/web_access/models.py]: No, you should save the mac-address for accountability of the trainees.

**Comment by radicalray** [comment]
LGTM!

**Pull request merged**